task_manager/task_manager.hh

```diff
@@ -0,0 +1,308 @@
+            task_id id;
+            std::string type;
+            task_state state = task_state::created;
+            db_clock::time_point start_time;
+            db_clock::time_point end_time;
+            std::string error;
+            int sequence_number = 0;        // A running sequence number of the task.
+            unsigned shard = 0;
+            std::string keyspace;
+            std::string table;
+            std::string entity;             // Additional entity specific for the given type of task.
+            std::string progress_units;     // A description of the units progress.
+        };
+
+        class impl {
+        protected:
+            status _status;
+            progress _progress;             // Reliable only for tasks with no descendants.
+            foreign_task_ptr _parent;
+            foreign_task_vector _children;
+            seastar::shared_promise<> _done;
+            module_ptr _module;
+        public:
+            impl(module_ptr module, task_id id, unsigned sequence_number, std::string keyspace, std::string table, std::string type, std::string entity, foreign_task_ptr&& parent) 
+                : _status({
+                    .id = id,
+                    .type = std::move(type),
+                    .state = task_state::created,
+                    .sequence_number = sequence_number,
+                    .shard = this_shard_id(),
+                    .keyspace = std::move(keyspace),
+                    .table = std::move(table),
+                    .entity = std::move(entity)
+                    })
+                , _parent(std::move(parent))
+                , _module(module)
+            {}
+
+            virtual future<progress> get_progress() const = 0;
+            virtual is_abortable is_abortable() const noexcept = 0;
+            virtual future<> abort() = 0;
+        protected:
+            virtual task_manager::task_ptr create_child(task_id id, task_manager::foreign_task_ptr&& parent = nullptr, unsigned sequence_number = 0, std::string keyspace = "", std::string table = "", std::string type = "", std::string entity = "") const = 0;
+            friend task;
+        };
+        using task_impl_ptr = std::unique_ptr<impl>;
+    protected:
+        task_impl_ptr _impl;
+    public:
+        task(task_impl_ptr&& impl) : _impl(std::move(impl)) {}
+
+        status& get_status() noexcept {
+            return _impl->_status;
+        }
+
+        task_id get_parent_id() const noexcept {
+            if (_impl->_parent) {
+                return _impl->_parent->_impl->_status.id;
+            }
+            return task_id::create_null_id();
+        }
+
+        foreign_task_ptr& get_parent() noexcept {
+            return _impl->_parent;
+        }
+
+        void set_type(std::string type) noexcept {
+            _impl->_status.type = std::move(type);
+        }
+
+        void set_total_progress(double progress) {
+            _impl->_progress.total = progress;
```

Comment:
Is it used?
With task::impl get_progress method I donxt think we need it.

Reply:
Currently it isn't used anywhere, though during task execution total progress needs to be set and progress must be reported somehow (for leaves).

Reply:
The question is whether progress should be pushed upwards or pulled when requested.
I'm inclined towards pulling it on demand rather than keeping the state in the "root" task.